### PR TITLE
Webpack3 Integration: Build and Dev support

### PR DESCRIPTION
**Pull request by zmaybury**
WeaveApp: 

- Build: `npm run build`
  - Will build the production distribution for WeaveApp and save to `dist` folder
- Dev: `npm run dev`
  - Will build a development version of WeaveApp and start a local dev server at `https://localhost:8080`. Local dev server will monitor source code for changes and will incrementally build and hot reload when changes are detected

WeaveJS:
- Build: `npm run build`
  - Will compile weavejs and then build the production distribution of WeaveApp (per above)
- Dev: `npm run dev`
  - Will compile weavejs and then build the development version of WeaveApp (per above) and start a local dev server (per above)

### CODE REVIEW

File: WeaveApp/resources/index.ejs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
Comment (zmaybury) [inline WeaveApp/resources/index.ejs:1]: Webpack allows templating for entry point html files, so in the future we can do things like add custom variables to this for custom titles, meta tags, scripts, etc.

File: WeaveApp/webpack.common.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
+const path = require('path');
```
Comment (zmaybury) [inline WeaveApp/webpack.common.js:1]: This is all the common webpack config required for either a development or production build.

File: WeaveApp/webpack.common.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            React: "React", react: "React", "window.react": "React", "window.React": "React"
+        }),
+    ],
+    output: {
+        filename: '[name].bundle.js',
+        path: path.resolve(__dirname, distFolder),
+    },
+    module: {
+        rules: [{
+            test: /\.tsx?$/,
```
Comment (zmaybury) [inline WeaveApp/webpack.common.js:58]: This puts all the typescipt files through both a typescript and babel translator, turning them into vanilla JS

File: WeaveApp/webpack.dev.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+const path = require('path');
```
Comment (zmaybury) [inline WeaveApp/webpack.dev.js:1]: This is all the configuration required to spin up the local development server.

File: WeaveApp/webpack.prod.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+const merge = require('webpack-merge');
```
Comment (zmaybury) [inline WeaveApp/webpack.prod.js:1]: This is the production configuration, comments inline.

File: WeaveApp/webpack.prod.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+const UglifyJSPlugin = require('uglifyjs-webpack-plugin');
+const Visualizer = require('webpack-visualizer-plugin');
+const common = require('./webpack.common.js');
+const webpack = require('webpack');
+
+module.exports = merge(common, {
+    devtool: 'source-map',
+    plugins: [
+        new UglifyJSPlugin({
```
Comment (zmaybury) [inline WeaveApp/webpack.prod.js:10]: This plugin uglify's and minify's the compiled code. It also creates a source map.

File: WeaveApp/webpack.prod.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+            sourceMap: true,
+        }),
+        new webpack.DefinePlugin({
```
Comment (zmaybury) [inline WeaveApp/webpack.prod.js:13]: This defines the node environment as production, which has a number of performance gains for libraries we use, specifically react.

File: WeaveApp/webpack.prod.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+            'process.env': {
+                'NODE_ENV': JSON.stringify('production'),
+            },
+        }),
+        new webpack.HashedModuleIdsPlugin(),
+        new Visualizer({
```
Comment (zmaybury) [inline WeaveApp/webpack.prod.js:19]: This plugin creates a static file, called `statistics.html`, that creates a visualization of our application and all its sub dependencies. This is especially useful for identifying dependencies that are particularly heavy, and may be targets for replacement by lighter alternatives.

**Comment by zmaybury** [comment]
@franckamayou @wdwalker : These are the weavejs-module files that I had in my `WeaveJS/lib` folder. If you add these files to your `WeaveJS/lib` folder and run the following commands, you should be up an running:
```
npm run compile-weavejs
cd WeaveAPP
npm install
npm run copy-libs
npm run dev
```
[weavejs-module.zip](https://github.com/WeaveTeam/WeaveJS/files/1331450/weavejs-module.zip)

**Review by franckamayou — APPROVED** [review]

**Pull request merged**